fairseq/models/transformer.py

```diff
@@ -397,15 +403,18 @@ def forward(self, src_tokens, src_lengths, cls_input=None, return_all_hiddens=Fa
             encoder_padding_mask = None
 
         encoder_states = [] if return_all_hiddens else None
+        encoder_attn = [] if return_all_attn else None
 
         # encoder layers
         for layer in self.layers:
             # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
             dropout_probability = random.uniform(0, 1)
             if not self.training or (dropout_probability > self.encoder_layerdrop):
-                x = layer(x, encoder_padding_mask)
+                x, attn = layer(x, encoder_padding_mask, need_head_weights=return_all_attn)
```

Comment:
The problem is that when `return_all_attn=False` then you only return `x`. You need to do something like:
```
layer_out = layer(x, encoder_padding_mask, need_head_weights=return_all_attn)
x = layer_out[0]
attn = layer_out[1] if len(layer_out) > 1 else None
```

Reply:
I'm confused because `attn` is always returned from EncoderLayer :

https://github.com/pytorch/fairseq/blob/3311af162f5806defd519c70688200171c22b867/fairseq/modules/transformer_layer.py#L118

(I did it like this to follow the DecoderLayer)

---

Also from the tests :

>x, attn = layer(x, encoder_padding_mask, need_head_weights=return_all_attn)
>ValueError: too many values to unpack (expected 2)

Meaning more than 2 values are returned ??

I couldn't run the tests locally, I think I should focus on this before anything else...

Thanks for the help ! I will come back to this PR after understanding the tests better !

Reply:
You can run the tests with `python setup.py test`